test/functional/treesitter/parser_spec.lua

```diff
@@ -959,4 +959,57 @@ int x = INT_MAX;
       )
     end)
   end)
+
+  it('option reload works', function()
+    eq(
+      true,
+      exec_lua('return vim.treesitter.get_parser(0, "lua") == vim.treesitter.get_parser(0, "lua")')
+    )
+    eq(
+      false,
+      exec_lua(
+        'return vim.treesitter.get_parser(0, "lua") == vim.treesitter.get_parser(0, "lua", {reload=true})'
+      )
+    )
+  end)
+
+  it('reloading parser refreshes queries', function()
+    local plug_dir = vim.uv.fs_mkdtemp(vim.fs.dirname(t.tmpname()) .. '/pluginXXXXXX')
+    local query_dir = plug_dir .. '/queries/lua'
+
+    exec_lua('vim.fn.mkdir(..., "p")', query_dir)
+
+    t.write_file(
+      query_dir .. '/injections.scm',
+      [[;; extends
+        (string 
+          content: _ @injection.content
+          (#lua-match? @injection.content "^TEST_INJ")
+          (#set! injection.language "c"))
+      ]]
+    )
+
+    insert([[
+      local s = "TEST_INJ test"
+    ]])
+
+    local langs = exec_lua(
+      [[
+      vim.treesitter.get_parser(0, "lua")
+      vim.opt.rtp:append(...)
+      collectgarbage() -- without it keeps cached queries
```

Comment:
@clason speaking about `vim.treesitter.query.get()`, I hit cache in tests, but never did when just running nvim, so I thought I misunderstood how caching suppose to work there.

I guess we need to invalidate query cache when getting new parser with `reload = true`?

Reply:
Yes, indeed, as reloading the queries is what you want in the first place. Possibly it's enough for this purpose if we allow invalidating the cache explicitly (maybe for now with a private function prefixed by `_`).

Reply:
I did not completely get what you mean by "invalidating the cache explicitly with a private function", could you elaborate, please?

Looked into `memoize` function and it is a closure over `cache`, and so cache is not accessible from outside. 

https://github.com/neovim/neovim/blob/4e5c633ed4871a948aff7338b793ac5f93484153/runtime/lua/vim/func/_memoize.lua#L41-L63

It can be converted to a table with `__call` that calls original function and some method like `_invalidate` attached, that clears cache (or specific key). And this would allow us to do something like this `vim.treesitter.query.get._invalidate_cache()`.

Or another option is to change what parameters used for cache, I tried this

```lua
M.get = function(lang, query_name)
  return getCachedQuery(lang, query_name, vim.o.rtp)
end
```

and `getCachedQuery` is original function, with one param for cache
```lua
local getCachedQuery = memoize('concat-3', function(lang, query_name, _rtp)
...
end)
```

Reply:
changed `memoize` could look something like this
```lua
return function(hash, fn, strong)
  -- ...
  local cache = {}
  local invalidate_cache = function(...)
    if #{ ... } == 0 then
      cache = {}
    else
      cache[hash(...)] = nil
    end
  end

  local wrapped = function(_self, ...)
    -- ...
  end

  return setmetatable({ _invalidate_cache = invalidate_cache }, { __call = wrapped })
end
```

Reply:
Settled for changing memozie function and added `treesitter.query._reset_cache` function, that calls _reset_cache internally, but allows simplier api for clearing all queries for lang: `query._reset_cache("lua")`.

Updated tests as well.